### PR TITLE
FSE: Fix navigation to and back from templates in Calypso

### LINKED ISSUE #35345: FSE: Template back button goes to page list, not parent page.
#### Steps to reproduce
0. Build and sync the FSE and iframe bridge server files to your sandbox. Also sandbox the api and widgets.wp.com.
1. With FSE enabled, open the page editor on your sandbox site.
2. Edit the header.
3. Click the back button. It will take you to the page list.

It should take you to the page you were originally editing.

Reported by @vindl when testing the FSE sync diff.

See also the original fix PR: https://github.com/Automattic/wp-calypso/pull/35005

Some original issues reporting this:
https://github.com/Automattic/wp-calypso/issues/34074
https://github.com/Automattic/wp-calypso/issues/34071

Note that this is NOT the same issue as https://github.com/Automattic/wp-calypso/issues/35105. (Unless investigation reveals a solution that solves both bugs 😁 )

### PR DESCRIPTION
**Pull request by noahtallen**
Because of a few changes to some other parts of FSE, we ended up breaking a few parts of navigation in wpcom including in the following situations:
- Clicking "Edit Header/Footer" would not take you to the Calypso URL
- Using the back button from the Template took you to the pages list instead of the page you were on.
- When you had an unsaved draft, clicking "Edit header/footer" would not take you to the calypso URL after saving, even after fixing the above issue.
- When you visited the Calypso template URL, the back button with the text didn't show up, just the "x".

This PR resolves all of those issues.

#### Changes proposed in this Pull Request
1. The first issue was caused by a change in when we show the edit buttons. Previously, we expected those buttons to be on the DOM all the time. Now that they are shown while focusing the template block, our code which replaced their hrefs could not find them. I changed that code to now look for those buttons and update their URLs when the template blocks are selected.
2. The second issue was caused by our calypso nav logic still using `wp_template_part` to determine if we should perform a certain type of navigation. I changed that to the new `wp_template`, as well as renamed all the functions and variables in the iframe which referenced template parts.
3. The third issue was caused by the code which autosaves the post when you navigate to edit a template. The autosave code, if the post was dirty, cancelled the default click event, saved the post, then manually set the window's location. To fix this, I used a ref to re-click the button after the save finishes. The effect of this is "pausing" the event until the save finishes, then "resuming it". It's needed because we manually override that button's href from the bridge server to use our calypso URLs, which are not available in FSE.
4. The fourth issue was caused by the parent page ID not being passed down as a query param into the iframe context.

I now see that this could have been broken into a few pieces, but it was all pretty intertwined as I was debugging it.

#### Setup instructions
This is a doozy! You'll need all three environments running to test this.

1. Pull this branch.
1. You'll want to sandbox a site for testing, `public-api.wordpress.com`, and `widgets.wp.com`.
2. You'll want to run calypso locally: `nvm use && npm i && npm start`, then go to `calypso.localhost:3000`
3. You'll need to build the bridge server: `nvm use && npx lerna run build --scope='@automattic/wpcom-block-editor'` then sync it to your sandbox.
4. You'll need to build the FSE code: `nvm use && npx lerna run dev --scope='@automattic/full-site-editing'` then sync it to your sandbox.

#### Testing instructions
Now that everything is running, 
1. Go to your sandbox site's pages list in the local calypso setup.
2. Click on an existing page. It should load and you should see the templates.
3. Once on the existing page, select the header and click "Edit Header". It should take you to the template editor for the header **and the URL should change to match the calypso template editor route** (`http://calypso.localhost:3000/block-editor/edit/wp_template/{sandbox-site}/{template-id}?fse_parent_post={parent-page-id}`)
3. Make sure the template editor page **uses the "Back to Page" button** from this Calypso URL
4. Click the "Back to Page" button. It should take you to the page you were just on.
5. Now make a few changes to the page without saving it.
6. Select the header and click "Edit Header". It should save the page, and then take you to the template editor for the header, and the URL should change.
7. Click the back button again and it should take you back to that page. And your new changes should still be there.
8. Now, go back to the pages list and create a new page. Select a template, but don't publish anything.
9. Select the header and click "Edit Header". It should create a draft for the page, save it, and then take you to the template editor for the header. Click the back button. It should take you back to the page you just created.
10. Also try these steps with the footer.

Fixes #35345, #35355

For followup:
- In wpadmin, "Back to Page" button takes you to the calypso flows even if you started in wpadmin flows. (https://github.com/Automattic/wp-calypso/issues/35105)

### CODE REVIEW

**Comment by noahtallen** [comment]
This is because the close button label and URL are both null in the close button override plugin when loading the template from the calypso route.

**Comment by gwwar** [comment]
Hmm @mattwiebe we may want to relax the script in D31490-code to allow overwriting to the current version

File: apps/wpcom-block-editor/src/calypso/iframe-bridge-server.js [hunk 1]
```diff
@@ -611,59 +611,58 @@ function openCustomizer( calypsoPort ) {
...
+
+		// When the template block becomes selected, we want to change
+		// the link of the "edit template" button.
+		const selectedBlock = select( 'core/editor' ).getSelectedBlock();
+
+		if ( selectedBlock && 'a8c/template' === selectedBlock.name ) {
+			const getImpendingLinks = setInterval( () => {
+				const editTemplateLink = document.querySelector(
+					'[data-type="a8c/template"] .template-block__overlay a'
+				);
```
Comment (gwwar) [inline apps/wpcom-block-editor/src/calypso/iframe-bridge-server.js]: Is there not a better way for us to pass this along?

On template, select we wait for the following element to render:
Screen Shot 2019-08-13 at 5 44 08 PM

Comment (gwwar) [reply apps/wpcom-block-editor/src/calypso/iframe-bridge-server.js]: 🤔 I guess we did this previously, so I won't block on this. I'm thinking we can pass along both of these links in a more robust way, as a follow up.

**Comment by gwwar** [comment]
@bsessions85 any idea who's owner on test-e2e-canary-woo? It seems odd that it's failing here. Changes should be unrelated.

**Comment by noahtallen** [comment]
Hmm canary woo seems to be failing a lot for me lately on PRs. Maybe similar to other failures folks have been seeing.

I definitely think there should be a better way to do this. I tried to not switch it up a ton since we're trying to fix it quickly! Maybe we could do it by overriding an URL with a filter like we do for the back button URL.

**Comment by vindl** [comment]
Started a diff to sync Block Editor changes in D31501-code.

**Pull request merged**